compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -2328,21 +2328,77 @@ public void visit(BLangXMLNS xmlnsNode) {
     }
 
     public void visit(BLangCompoundAssignment compoundAssignment) {
-        BLangAssignment assignStmt = (BLangAssignment) TreeBuilder.createAssignmentNode();
-        assignStmt.pos = compoundAssignment.pos;
-        BLangVariableReference varRef = compoundAssignment.varRef;
+        // If compound Assignment is an index based expression such as a[f(1, foo)][3][2] += y,
+        // should return a block statement which is equivalent to
+        // var $temp1$ = 2;
+        // var $temp2$ = 3;
+        // var $temp3$ = a[f(1, foo)];
+        // a[$temp3$][$temp2$][$temp1$] = a[$temp3$][$temp2$][$temp1$] + y;
+        if (compoundAssignment.varRef.getKind() == NodeKind.INDEX_BASED_ACCESS_EXPR) {
+            int indexExprNumber = 0;
+            List<BLangStatement> statements = new ArrayList<>();
+            List<BLangSimpleVarRef> varRef = new ArrayList<>();
+            List<BType> type = new ArrayList<>();
+
+            BLangIndexBasedAccess indexBasedAccess = (BLangIndexBasedAccess) compoundAssignment.varRef;
+            // Extract the index Expressions from compound assignment and create variable definitions. ex:
+            // var $temp1$ = 2;
+            // var $temp2$ = 3;
+            // var $temp3$ = a[f(1, foo)];
+            do {
+                indexExprNumber += 1;
+                BLangSimpleVariableDef tempIndexVarDef = createVarDef("$temp" + indexExprNumber + "$",
+                        indexBasedAccess.indexExpr.type, indexBasedAccess.indexExpr, compoundAssignment.pos);
+                BLangSimpleVarRef tempVarRef = ASTBuilderUtil.createVariableRef(tempIndexVarDef.pos,
+                        tempIndexVarDef.var.symbol);
+                statements.add(tempIndexVarDef);
+                varRef.add(tempVarRef);
+                type.add(indexBasedAccess.type);
+
+                if (indexBasedAccess.expr.getKind() == NodeKind.INDEX_BASED_ACCESS_EXPR) {
+                    indexBasedAccess = (BLangIndexBasedAccess) indexBasedAccess.expr;
+                    continue;
+                }
+                break;
+            } while (true);
```

Comment:
Is there a workaround for us to not use `while(true)`?
How about something as follows?
```
do {
// do the casting somewhere here

} while(indexBasedAccess.expr.getKind() == NodeKind.INDEX_BASED_ACCESS_EXPR);
```

@MaryamZi  thoughts? is it okay using `while(true)` here?

Reply:
Yes, the condition to check is whether the current var ref is a index based access expr. No need for a do-while.

Reply:
Instead of L2358-L2362 can we do something like
- `BLangExpression expr = compoundAssignment.varRef;` at the very beginning 
- `expr = compoundAssignment.varRef;` instead of L2358-L2362
- update the check in while to `while (expr.getKind() == NodeKind.INDEX_BASED_ACCESS_EXPR);`

Reply:
If we change the code according to the above suggestion, the last expression does not go through the logic.
ex: Let's consider the first expression is `c[f(1, foo)][1][2]`. After two iterations, according to this `expr = compoundAssignment.varRef;` the expression becomes `c[f(1, foo)]`. According to the check inside the **while**, execution exits from the loop and `c[f(1, foo)]` does not execute. But `c[f(1, foo)]` should go through the loop.

Reply:
Please ignore the above comment. Misread the suggestion. Changed the code according to the suggestion.